include/arch/arc/v2/irq.h

```diff
@@ -40,18 +40,20 @@ extern void z_irq_priority_set(unsigned int irq, unsigned int prio,
 extern void _isr_wrapper(void);
 extern void z_irq_spurious(const void *unused);
 
+
+/* lowest priority */
+#define ARCH_DEFAULT_IRQ_PRIORITY	(CONFIG_NUM_IRQ_PRIO_LEVELS - 1)
+
 /* Z_ISR_DECLARE will populate the .intList section with the interrupt's
  * parameters, which will then be used by gen_irq_tables.py to create
- * the vector table and the software ISR table. This is all done at
- * build-time.
- *
- * We additionally set the priority in the interrupt controller at
- * runtime.
+ * the interrupt vector table, interrupt priority table and the software ISR table.
+ * This is all done at build-time.
  */
 #define ARCH_IRQ_CONNECT(irq_p, priority_p, isr_p, isr_param_p, flags_p) \
 { \
-	Z_ISR_DECLARE(irq_p, 0, isr_p, isr_param_p); \
-	z_irq_priority_set(irq_p, priority_p, flags_p); \
```

Comment:
Hi Ruud,

please don't forget that we are moving info from one place to another :)

I've done some measurements:

---
If we actually use interrupt lines (if we connect handlers to them) there is even a image size decrease:

Setup with IRQ used, hsdk board (CONFIG_NUM_IRQS=88), without changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32728 B      1280 MB      0.00%
```

Setup with IRQ used, hsdk board (CONFIG_NUM_IRQS=88), with changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32124 B      1280 MB      0.00%
```

Setup with IRQ used, nsim_hs board (CONFIG_NUM_IRQS=30), without changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9896 B         1 MB      0.94%
            DCCM:        3672 B         1 MB      0.35%
```

Setup with IRQ used, nsim_hs board (CONFIG_NUM_IRQS=30), with changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9780 B         1 MB      0.93%
            DCCM:        3672 B         1 MB      0.35%
```

----

However if we have a setup with a lot of interrupts which are **not used**, there is a small penalty:

Setup with IRQ unused, nsim_hs board (CONFIG_NUM_IRQS=30), without changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9768 B         1 MB      0.93%
            DCCM:        3672 B         1 MB      0.35%
```

Setup with IRQ unused, nsim_hs board (CONFIG_NUM_IRQS=30), with changes:
```
Memory region         Used Size  Region Size  %age Used
            ICCM:        9776 B         1 MB      0.93%
            DCCM:        3672 B         1 MB      0.35%
```

Setup with IRQ unused, hsdk board (CONFIG_NUM_IRQS=88), without changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32088 B      1280 MB      0.00%
```


Setup with IRQ unused, hsdk board (CONFIG_NUM_IRQS=88), with changes:
```
Memory region         Used Size  Region Size  %age Used
            SRAM:       32120 B      1280 MB      0.00%
```

The good news is that we can strip interrupt table (and interrupt priority table) if we don't use some of the interrupts.

There are another possible optimizations:
- we can disable interrupt priority table generation if we have only one interrupt priority level.
- we can introduce interrupt priority table packing (i.e on ARC we have 16 interrupt priorities levels, so is enough to have 4bits to store them)

However, I guess it would be better to have these optimizations in a separate PR.

Reply:
Great to see the numbers, indeed the trade-off is between constant data and code. The increase without (many) ISRs installed is very small, let's avoid more #ifdef's. Other optimizations would (like you suggest) have more impact in that case.